pkg/authorization/authorizer/scope/authorizer.go

```diff
@@ -0,0 +1,71 @@
+package scope
+
+import (
+	"fmt"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/util/errors"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	defaultauthorizer "github.com/openshift/origin/pkg/authorization/authorizer"
+	"github.com/openshift/origin/pkg/authorization/rulevalidation"
+)
+
+type scopeAuthorizer struct {
+	delegate            defaultauthorizer.Authorizer
+	clusterPolicyGetter rulevalidation.ClusterPolicyGetter
+}
+
+func NewAuthorizer(delegate defaultauthorizer.Authorizer, clusterPolicyGetter rulevalidation.ClusterPolicyGetter) defaultauthorizer.Authorizer {
+	return &scopeAuthorizer{delegate: delegate, clusterPolicyGetter: clusterPolicyGetter}
+}
+
+func (a *scopeAuthorizer) Authorize(ctx kapi.Context, passedAttributes defaultauthorizer.AuthorizationAttributes) (bool, string, error) {
+	user, exists := kapi.UserFrom(ctx)
+	if !exists {
+		return false, "", fmt.Errorf("no user")
+	}
+
+	uncastScopes := user.GetExtra()["scopes"]
+	scopes, ok := uncastScopes.([]string)
+	if !ok || len(scopes) == 0 {
+		return a.delegate.Authorize(ctx, passedAttributes)
+	}
+
+	nonFatalErrors := []error{}
+
+	namespace, _ := kapi.NamespaceFrom(ctx)
+	// scopeResolutionErrors aren't fatal.  If any of the scopes we find allow this, then the overall scope limits allow it
+	rules, err := ScopesToRules(scopes, namespace, a.clusterPolicyGetter)
```

Comment:
checking attributes against the rules as we build them and short-circuiting seems like a good future optimization